bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/KeypadMessage.java

```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.alarmdecoder.internal.protocol;
+
+import java.util.List;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * The {@link KeypadMessage} class represents a parsed keypad (KPM) message.
+ * Based partly on code from the OH1 alarmdecoder binding by Bernd Pfrommer.
+ *
+ * @author Bob Adair - Initial contribution
+ */
+@NonNullByDefault
+public class KeypadMessage extends ADMessage {
+
+    // Example: [00110011000000003A--],010,[f70700000010808c18020000000000],"ARMED ***STAY** ZONE BYPASSED "
+
+    public static final int BIT_READY = 17;
+    public static final int BIT_ARMEDAWAY = 16;
+    public static final int BIT_ARMEDHOME = 15;
+    public static final int BIT_BACKLIGHT = 14;
+    public static final int BIT_PRORGAM = 13;
+    public static final int BIT_BYPASSED = 9;
+    public static final int BIT_ACPOWER = 8;
+    public static final int BIT_CHIME = 7;
+    public static final int BIT_ALARMOCCURRED = 6;
+    public static final int BIT_ALARM = 5;
+    public static final int BIT_LOWBAT = 4;
+    public static final int BIT_DELAYOFF = 3;
+    public static final int BIT_FIRE = 2;
+    public static final int BIT_SYSFAULT = 1;
+    public static final int BIT_PERIMETER = 0;
+
+    public final String bitField;
+    public final int numericCode;
+    public final String rawData;
+    public final String alphaMessage;
+    public final int nbeeps;
+    public final int status;
+
+    private final int upper;
+    private final int lower;
+
+    public KeypadMessage(String message) throws IllegalArgumentException {
+        super(message);
+        List<String> parts = splitMsg(message);
+
+        if (parts.size() != 4) {
+            throw new IllegalArgumentException("Invalid number of parts in keypad message");
+        }
+        if (parts.get(0).length() != 22) {
+            throw new IllegalArgumentException("Invalid field length in keypad message");
+        }
+
+        bitField = parts.get(0);
+        rawData = parts.get(2);
+        alphaMessage = parts.get(3).replaceAll("^\"|\"$", "");
+
+        try {
+            int numeric = 0;
+            try {
+                numeric = Integer.parseInt(parts.get(1));
+            } catch (NumberFormatException e) {
+                numeric = Integer.parseInt(parts.get(1), 16);
+            }
```

Comment:
I believe so. This comes from the original V1 binding code, so it has been running happily lo these many years. :-) I expect that in the rare instances this field contains a hex number, it will be prefixed with 0x so that the first parseInt will reliably throw an exception.
As a side note... who comes up with a protocol where an integer represented as text may be either decimal or hex? Sigh.

Reply:
> As a side note... who comes up with a protocol where an integer represented as text may be either decimal or hex? Sigh.

Probably not one person, but a group of people that don't check or correct each-other's work managed by a manager that doesn't care.

Reply:
LOL. That's the sort of thing that slips through as a team approaches a release deadline.